node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,452 @@
+
+/// Implements the `select_disputes` function which selects dispute votes which should
+/// be sent to the Runtime.
+///
+/// # How the prioritization works
+///
+/// Generally speaking disputes can be described as:
+///   * Active vs Concluded
+///   * Known vs Unknown onchain
+///   * Local vs Onchain
+///   * Runtime knows about enough votes to conclude the dispute vs
+///     Runtime needs more votes to conclude the dispute.
+///
+/// Provisioner fetches all disputes from `dispute-coordinator` and separates them in five partitions
+/// (listed in order of importance) based on this criteria:
+///
+///   1. Active local disputes, also known onchain, without enough votes to conclude onchain.
+///   2. Active local disputes, unknown onchain.
+///   3. Active local disputes, also known onchain, with enough votes to conclude onchain.
+///   4. Concluded local disputes, unknown onchain.
+///   5. Concluded local disputes, known onchain.
+///
+/// `disputes-coordinator` first sends all disputes in group 1, then if there is more space - group 2 and
+/// then group 3. Group 4 is supposed to be empty but it is sent nevertheless. Group 5 is completely dropped.
+///
+/// Or a bit more visually:
+/// 	- Active
+/// 		- Known onchain
+/// 			- Not enough votes to conclude (PARTITION 1)
+/// 			- With enough votes to conclude	(PARTITION 3)
+/// 		- Unknown onchain (PARTITION 2)
+/// 	- Concluded
+/// 		- Known onchain (THESE ARE DROPPED)
+/// 		- Unknown onchain (PARTITION 4)
+///
+/// # Votes selection
+///
+/// Besides the prioritization described above the votes in each partition are filtered too. Provisioner
+/// fetches all onchain votes and filters them out from all partitions. As a result the Runtime receives
+/// only fresh votes (votes it didn't know about).
+///
+/// # How the onchain votes are fetched
+///
+/// The logic outlined above relies on `RuntimeApiRequest::StagingDisputes` message from the Runtime staging API.
+/// If the staging API is not enabled - the same logic is executed with empty onchain votes set. Effectively this
+/// means that all disputes are partitioned in groups 2 or 4 and all votes are sent to the Runtime.
+
+/// The maximum number of disputes Provisioner will include in the inherent data.
+/// Serves as a protection not to flood the Runtime with excessive data.
+// The magic numbers below are: `estimated validators count` * `estimated disputes per validator`
+pub const MAX_DISPUTE_VOTES_FORWARDED_TO_RUNTIME: usize = 200 * 1_000;
+pub async fn select_disputes<Sender>(
+	sender: &mut Sender,
+	metrics: &metrics::Metrics,
+	leaf: &ActivatedLeaf,
+) -> Result<MultiDisputeStatementSet, Error>
+where
+	Sender: overseer::ProvisionerSenderTrait,
+{
+	gum::trace!(target: LOG_TARGET, ?leaf, "Selecting disputes for inherent data");
+
+	// Fetch the onchain disputes. We'll do a prioritization based on them.
+	let onchain = match get_onchain_disputes(sender, leaf.hash.clone()).await {
+		Ok(r) => r,
+		Err(e) => {
+			gum::debug!(
```

Comment:
I agree. I'll add more verbose error logging.
I wanted to avoid heavy match expressions here but the current version of the code will hide execution errors as you mentioned.

Another solution is to add the error 'at the source' (inside the module) but I think it's better to give the caller information what happened instead of hiding errors.